File: android/app/src/main/java/com/converse/dev/Datatypes.kt
package com.converse.dev

import kotlin.math.roundToInt

class NotificationData(val message: String, val timestampNs: String, val contentTopic: String, val account: String)
class ConversationDictData(val shortAddress: String? = null, val title: String? = null)
class SavedNotificationMessage(val topic: String, val content: String, val senderAddress: String, val sent: Long, val id: String, val contentType: String, val account: String? = null, val referencedMessageId: String? = null)
class ConversationContext(val conversationId: String, val metadata: Map<String, Any>)
class Accounts(val currentAccount: String, val accounts: Array<String>, val databaseId: Map<String, String>? = null)
class AccountsStore(val state: Accounts, val version: Int)
class EnsName(val name: String, val displayName: String? = null, val isPrimary: Boolean? = false)
class ConverseUserName(val name: String, val displayName: String? = null, val isPrimary: Boolean? = false)
class ProfileSocials(val ensNames: Array<EnsName>? = null, val userNames: Array<ConverseUserName>? = null)
class Profile(val updatedAt: Long, val socials: ProfileSocials)
class Profiles(val profiles: Map<String, Profile>? = null)
class ProfilesStore(val state: Profiles, val version: Int)
class SavedNotificationConversation(val topic: String, val peerAddress: String, val createdAt: Long, val context: ConversationContext?, val account: String? = null, spamScore: Double?) {
  // Whenever spamScore is set, round it to two decimal to ensure it fits as a 2-digit float in the database
  var spamScore: Double? = spamScore
    set(value) {
      field = value?.let { (it * 100).roundToInt().toDouble() / 100 }
    }
}

File: ios/ConverseNotificationExtension/Datatypes.swift
//
//  Datatypes.swift
//  ConverseNotificationExtension
//
//  Created by Noe Malzieu on 28/08/2023.
//

import Foundation

struct SavedNotificationMessage: Codable {
  var topic: String
  var content: String
  var senderAddress: String
  var sent: Int
  var id: String
  var contentType: String
  var account: String
  var referencedMessageId: String?
}

struct ConversationContext: Codable {
  var conversationId: String
  var metadata: Dictionary<String, String>
}

struct SavedNotificationConversation: Codable {
  var topic: String
  var peerAddress: String?
  var createdAt: Int
  var context: ConversationContext?
  var account: String
  var isGroup: Bool = false
  var groupMembers: [String?]?
  var groupPermissionLevel: String?
  var groupName: String?
  var spamScore: Double? {
    didSet {
      if let value = spamScore {
        // Whenever spamScore is set, round it to two decimal to ensure it fits as a 2-digit float in the database
        spamScore = (value * 100).rounded() / 100
      }
    }
  }
}

struct Accounts: Codable {
  var currentAccount: String
  var accounts: [String]
  var databaseId: Dictionary<String, String>?
}

struct AccountsStore: Codable {
  var state: Accounts;
  var version: Int
}

struct EnsName: Codable {
  var name: String;
  var displayName: String?;
  var isPrimary: Bool?;
  var avatar: String?;
}


struct ConverseUserName: Codable {
  var name: String;
  var displayName: String?;
  var isPrimary: Bool?;
  var avatar: String?;
}

struct ProfileSocials: Codable {
  var ensNames: [EnsName]?;
  var userNames: [ConverseUserName]?;
}

struct Profile: Codable {
  var updatedAt: Int;
  var socials: ProfileSocials;
}
struct Profiles: Codable {
  var profiles: Dictionary<String, Profile>?
}

struct ProfilesStore: Codable {
  var state: Profiles;
  var version: Int
}
